File: src/main/java/org/tdf/lotusvm/types/GlobalType.java
package org.tdf.lotusvm.types;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import org.tdf.lotusvm.common.BytesReader;

/**
 * Global types are encoded by their value type and a flag for their mutability.
 */
@AllArgsConstructor
@Builder
@Getter
public class GlobalType {
    private ValueType valueType;
    private boolean mutable; // true var ,false const

    static GlobalType readFrom(BytesReader reader){
        ValueType valueType = ValueType.readFrom(reader);
        boolean mutable = reader.read() != 0;
        return GlobalType.builder().valueType(valueType).mutable(mutable).build();
    }
}

File: src/main/java/org/tdf/lotusvm/types/GlobalType.kt
package org.tdf.lotusvm.types

import org.tdf.lotusvm.common.BytesReader
import org.tdf.lotusvm.common.ObjectReader

/**
 * Global types are encoded by their value type and a flag for their mutability.
 */
data class GlobalType(
    val valueType: ValueType, // true var ,false const
    val isMutable: Boolean
) {

    companion object : ObjectReader<GlobalType> {
        override fun readFrom(reader: BytesReader): GlobalType {
            val valueType = ValueType.readFrom(reader)
            val mutable = reader.read() != 0
            return GlobalType(valueType, mutable)
        }
    }
}